### DISCUSSION TITLE
What is the use-case for synchronous fetcher?

### DISCUSSION THREAD

**Discussion by koba04** [vercel/swr#917] [General] [answered]
According to the type definition and implementation, the fetcher function seems to support a non-promise function.
What is the use-case for the synchronous fetcher?
I've seen synchronous fetchers in tests; Is it only for testing?
If there is no use-case, could I drop the synchronous fetcher from the type definition?

- Type Definition

https://github.com/vercel/swr/blob/a9c10f74963bbe3b566201ef9f43a2242440f0f1/src/types.ts#L1

- A test

https://github.com/vercel/swr/blob/a9c10f74963bbe3b566201ef9f43a2242440f0f1/test/use-swr-suspense.test.tsx#L81-L95

- Implementation checking if the value is a promise

https://github.com/vercel/swr/blob/a9c10f74963bbe3b566201ef9f43a2242440f0f1/src/use-swr.ts#L779-L785

**Comment by huozhi** [comment, marked as answer]
In my understanding, swr can be used as a state managment library for various cases, not only for handling network requests. The `revalidate` action could mean "retrieve state from data source", which can be either sync or async.

Imagine we're building a todo list app, swr hooks can still be a great helper to access and cache states of todo items.

```js
// getTodoItems can be a function reading todos from local states
const {data:todoItems} = useSWR('todos', getTodoItems)
```

here's another example in swr codebase https://github.com/vercel/swr/tree/master/examples/storage-tab-sync

**Reply by koba04** [reply]
Thank you! I didn't know the example.
I was confused about the `CONCURRENT_PROMISES` variable might store a non-promise value, so it would be nice if I can refactor the code, but I understood it's intentional. Thank you!

**Reply by morgs32** [reply]
It would be great if `useSWR` would immediately return data when the fetcher is synchronous. It seems to always return undefined. Any way to do that besides passing default data to SWRConfig??
